cirq/protocols/apply_unitary_to_tensor.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Union, Sequence, TypeVar
+
+import numpy as np
+from typing_extensions import Protocol
+
+from cirq import linalg
+from cirq.protocols.unitary import unitary
+
+
+# This is a special indicator value used by the apply_unitary_to_tensor method
+# to determine whether or not the caller provided a 'default' argument. It must
+# be of type np.ndarray to ensure the method has the correct type signature in
+# that case. It is checked for using `is`, so it won't have a false positive if
+# the user provides a different np.array([]) value.
+RaiseTypeErrorIfNotProvided = np.array([])  # type: np.ndarray
+
+TDefault = TypeVar('TDefault')
+
+
+class SupportsApplyUnitaryToTensor(Protocol):
+    """An object that can be efficiently left-multiplied into tensors."""
+
+    def _apply_unitary_to_tensor_(self,
+                                  target_tensor: np.ndarray,
+                                  available_buffer: np.ndarray,
+                                  axes: Sequence[int],
+                                  ) -> Union[np.ndarray, type(NotImplemented)]:
+        """Left-multiplies a unitary effect onto a tensor with good performance.
+
+        This method is given both the target tensor and workspace of the same
+        shape and dtype. The method then either performs inline modifications of
+        the target tensor and returns it, or writes its output into the
+        workspace tensor and returns that. This signature makes it possible to
+        write specialized simulation methods that run without performing large
+        allocations, significantly increasing simulation performance.
+
+        Args:
+            target_tensor: The input tensor that needs to be left-multiplied by
+                the unitary effect of the receiving object. The tensor will
+                have the shape (2, 2, 2, ..., 2).
```

Comment:
doc the rank of this tensor (since you say above you can apply it to a unitary as well as a state I assume you mean by this that you can apply to some arbitrary subset of indices)

Reply:
Acknowledged but I think it already describes that? The shape has to be (2, 2, 2, 2, ..., 2) and the axes parameter identifies the indices of some of those 2s.